sdk/cosmos/azure_data_cosmos/src/query.rs

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+use serde::Serialize;
+
+/// Represents a Cosmos DB Query, with optional parameters.
+///
+/// # Examples
+///
+/// Create a query using [`Query::from()`], and use  [`Query::with_parameter()`] to add parameters to it as needed.
+///
+/// ```rust
+/// # use azure_data_cosmos::{Query,NullValue};
+/// let query = Query::from("SELECT * FROM c WHERE c.id = @customer_id")
+///     .with_parameter("@customer_id", 42).unwrap();
+/// ```
+///
+/// # Specifying Parameters
+///
+/// Any JSON-serializable value, including the [`NullValue`](crate::NullValue) marker value, can be used as a parameter.
+/// The [`Query::with_parameter()`] method accepts any type that implements [`serde::Serialize`] as a value.
+/// Because the type needs to be serialized in order to be sent as a query parameter, the [`Query::with_parameter()`] method is fallible and may return [`Result::Err`].
+///
+/// ```rust
+/// # use azure_data_cosmos::{Query, NullValue};
+/// let query = Query::from("SELECT * FROM c WHERE ...")
+///     .with_parameter("@customer_id", 42).unwrap()
+///     .with_parameter("@customer_name", "Contoso").unwrap()
+///     .with_parameter("@is_active", true).unwrap()
+///     .with_parameter("@offer_code", NullValue).unwrap();
+/// ```
+///
+/// This includes arrays and objects, if they implement [`serde::Serialize`]:
+///
+/// ```rust
+/// # use azure_data_cosmos::{Query, NullValue};
+/// #[derive(serde::Serialize)]
+/// struct CustomerInfo {
+///     id: u64,
+///     name: String
+/// }
+/// let query = Query::from("SELECT * FROM c WHERE ...")
+///     .with_parameter("@customer_info", CustomerInfo { id: 42, name: "Contoso".into() }).unwrap();
+/// ```
+#[derive(Debug, Serialize)]
+pub struct Query {
+    query: String,
+
+    #[serde(skip_serializing_if = "Vec::is_empty")] // Don't serialize an empty array.
+    parameters: Vec<QueryParameter>,
+}
+
+impl Query {
+    /// Creates a new [`Query`] with the same text and the same parameters, with one additional parameter added.
```

Comment:
Technically not a new Query, right?  Same query that has been mutated?

Reply:
Technically, you are correct ;). However, I think it's best to keep this as-is. The method consumes `self` by value, so the original query is moved into this method and no longer available to the caller. The fact that what we _actually_ do is take that object and add a parameter to it is mostly an implementation detail at that point. We could just as easily be creating a brand new one.

I think there's probably a rephrase here that could be clearer though, since the nuances there aren't super important to the caller.